### ISSUE TITLE
Brainstorming the policy of deletion of local mod folders to ensure a clean reinstall 

### ISSUE THREAD

**Issue by solarsootysmudge** [KSP-CKAN/CKAN#1610]
This issue raised in reaction to comments on #1607. Raised a new issue because this is a different approach to dealing installation problems from the perspective of an end user. Plus I don't feel this helps the team implement #1607 and is off topic. None the less it is worthy of further review.  

Any improvement of Metadata checking is welcomed and I really like it. If there is something that can alert us when something significant in a mod's structure has changed locally. In some cases mods create local configuration files post-install that CKAN doesn't know about.

As an end user I see most problems arising when you uninstall such mod. You may think that you have uninstalled mod, but since there is files inside mod folder that was not handled by CKAN, whole mod folder still remains inside gamedata, but only partly, with those additional config files generated after first install. It is also possible that mod author changes sub folder structure which can stop old files from being replaced correctly. 

A more robust method for checking of Metadata is brilliant. However there is always the danger that an end user can report an error because they accidently hold on to out dated locally held files. Such users then have a habit of complaining to the mod author because the current CKAN download appears broken. This can understandably cause some resentment on the part of the mod author. 

It might be better that we give the option to forcibly remove all files including those that CKAN did not create. Through removal of the whole folder. However this may cause other concerns which should be investigated. 

TL;DR. Good metadata checking is wonderful. However end users must at least help by manually checking for files inside the mod folder that where not handled directly by CKAN. It is always currently best practice to remove the whole mod folder prior to reinstalling a new version of some mods. Perhaps we need to give an option to automate this.

**Comment by plague006** [comment]
Ref https://github.com/KSP-CKAN/CKAN/issues/700.

**Comment by nobodyhasthis** [comment]
The main concern for CKAN development is to not to delete any user custom settings for a particular mode as stated by @pjf in #653. However I still believe getting potentially harmful files out of the gamedata folder is doing this is the lesser of two evils. 

Some compromise might be to instead of deleting folder, CKAN move that folder to another location. Say for example:  <KSP_install>\CKAN\ModBackups\ 

This purges the mod from the gamedata folder whilst giving the user to a chance to recover stuff. I am saying that even after giving the option to purge we can still let a user change his mind. That is being fair to both end users and mod authors.

**Comment by politas** [comment]
I think it is worth having CKAN distinguish between "deleting a mod because upgrading to a new version" and "deleting a mod because the user requested it be gone". In the latter case, maybe the user _wants_ the configuration to go too.

**Comment by nobodyhasthis** [comment]
I can only judge the situation based on KSP forum feedback. Where there is evidence that end users are unaware that some mods will leave behind files. Even after CKAN uninstalls them. So when a mod author specifically asks for a completely clean update. Users just download the update and then complain that it is broken. Never fully understanding that they have made a mistake. 

The correct action was to remove the mod, plus the whole folder. Then reinstall the mod. Although this may seem a bit drastic. There has be circumstances in the past where this is the only way to successfully debug a mod. Yet is not naturally intuitive to end users. Who are in general not technically competent enough to see what could be wrong in the gamedata folder. I say that as diplomatically as possible because I am one of them. 

Also in the process of forcing such an update. If I do delete my own personal mod configuration files. I can't blame the CKAN development team or the mod authors. I should have had the common sense to make a backup.

Ref #653 for reasons not do it and why it needs a GUI input. #327 to safe guard against against user errors here.

**Comment by politas** [comment]
The trouble is, mods are not always in neat encapsulated folders. Many mod authors create an author folder that they place all their mods in, or share some folders between different mods, and package partial versions of other mods. If we've installed Firespitter-core as a dependency for mod X, and then remove Mod X, do we delete the Firespitter directory? There just isn't clear ownership of folders by specific mods.

**Comment by plague006** [comment]
#1507 has a bit of worthwhile reading on this topic.

**Comment by nobodyhasthis** [comment]
I could be wrong but I also think that that everything on file overwrites needs to be studied as well. Before tackling this issue. https://github.com/KSP-CKAN/CKAN/milestones/File%20Overwrites. 

There is a good exploration of scenarios where mod dependencies are removed or overwritten. In cases where the relationship is deeper that one level. Plus the methods required to detect mods that have been removed.

**Comment by mathuin** [comment]
My comment in #1507 was not completely addressed.

> Wouldn't it be better for a mod's metadata to include a reference to files and/or directories that it creates? That would allow for the deletion of said content at the time the mod is removed via CKAN.

I'm not talking about items that a mod **installs**.  I'm talking about files that a mod **creates** while running -- configuration files and the like.

**Comment by plague006** [comment]
> I'm talking about files that a mod creates while running -- configuration files and the like.

How would we get that information and how would we verify that information remains true with every new version of a mod?

**Comment by mathuin** [comment]
The information would be supplied by the mod author, or failing that, the person who volunteers to submit the mod to CKAN on behalf of the author.

**Comment by plague006** [comment]
Authors and agents working on their behalf represent roughly 15-30% of the indexed mods. What would you propose for the other 70-85%?

**Comment by mathuin** [comment]
In *my* ideal world, the ratio would be the other way around -- more authors being involved than not -- but it would be far more useful to have metadata about the mod's post-installation behavior.

**Comment by nobodyhasthis** [comment]
Well actually I raised the issue because a mod update broke. Despite the mod author also being the metadata maintainer. After a big overhaul of the mod it was necessary to change the mods file structure. The metadata got fixed but old files got left behind from the previous install. 

There is also a future case where Mod A affects Mod B. Mod B post-installation behavior has changed. We want to remove all traces of mod A. What happens to Mod B?

Making the metadata more complicated to update does not seem like a good solution. Next to asking end users to clean out the old files before adding the affected mod back in. 

Remove Mod A+B. Add Mod B back in.

I am not too keen to add something to the metadata workload. If fact I started this issue up to ask about solutions that didn't use extra metadata to track changes. Especially since the problems presented can avoided using the current inbuilt CKAN features and metadata format.  

Currently the workaround is to uninstall affected mods. Then export installed mods via .ckan. Then import them back into to a virgin gamedata folder. That checks dependency and version numbers on everything.  It also ensures that Module Manager is not trying to work with corrupted folders left behind after a bad unistall. Some problems like #978 and #327 are avoided too. All without asking more from Authors and agents working to keep the metadata straight. 

However that is not naturally intuitive to end users. They tend to take the problem into the forums and we all know how that can work out. There also a half dozen reasons why this is a bad workaround to introduce as an automated function. Which is covered in the issue reports referenced above.  So far it is bit of a Catch-22 here.

**Comment by mathuin** [comment]
Actually, that raises an interesting question.

Can that workaround be automated with Jenkins or some other CI tool?  Take a virgin install of KSP, install the mod and its dependencies, check the file system, somehow start the game enough for ModuleManager and friends to run, check the file system, compare the results?

This obviously wouldn't help with the specific case of mod X changing its behavior during an upgrade, but would it help at all in the general case?

**Comment by nobodyhasthis** [comment]
@mathuin Honest answer here is. I do not know enough to answer your question. All I can do is give some feedback from myself and other end users. In the hope that it helps.

My guess, and this is a huge guess. Is that is this not a good solution at the point where the game needs to actually run to verify the data. It is bit more simple that. I honestly think if you copied the gamedata folder. You would have a back up to 'roll back' to. It's state would be returned to point before CKAN touched it. We would not need any sophisticated file system scanning. I think people just want all traces gone. Who cares what the changes where. 

However even doing that. Is extra work. For a problem that affects a tiny number of users. Relative to some other problems. This is a really low priority problem. One that could just be fixed by better user documentation. People just need to know that export function. It is available right now. It does fix a lot of problems. No need for a lot of extra code.

I feel the best solution here is to do nothing right now. If we can improve the bug reporting system. To the point where users get sent to a self help page. The problem might get removed. Now I know that is horribly simplistic. However asking somebody to code a better solution kind of feels like a waste of precious resources.

**Comment by politas** [comment]
#1873 should fix most of the problems referenced here.

**Comment by nobodyhasthis** [comment]
#1730 is now merged. Also reduces the odds seeing some of the problems presented here.

**Comment by solarsootysmudge** [comment]
Good stuff and now the edge cases can be solved by following the workaround above. Enhanced and full documented pages recently added to user guide complete this.
 https://github.com/KSP-CKAN/CKAN/wiki/Clean-and-reinstall-process
